src/dfm/cleaning/quality_filter.py

```diff
@@ -0,0 +1,697 @@
+"""
+Danish implementation of quality filter and repetitious text filter described in [1]
+
+Authors:
+    Kenneth C. Enevoldsen
+    Kasper Junge
+    Malte Højmark-Bertelsen
+
+References:
+    [1] Rae, J. W., Borgeaud, S., Cai, T., Millican, K., Hoffmann, J., Song, F.,
+    Aslanides, J., Henderson, S., Ring, R., Young, S., Rutherford, E., Hennigan,
+    T., Menick, J., Cassirer, A., Powell, R., Driessche, G. van den, Hendricks,
+    L. A., Rauh, M., Huang, P.-S., … Irving, G. (2021).
+    Scaling Language Models: Methods, Analysis & Insights from Training Gopher.
+    https://arxiv.org/abs/2112.11446v2
+"""
+
+from typing import Iterable, Optional, Set, Tuple, List, Callable
+from collections import Counter, defaultdict
+from functools import partial
+
+import spacy
+from spacy.tokens import Doc
+
+
+def duplicate_chr_fraction_getter(doc: Doc, attr: str) -> float:
+    """Calculate the character fraction of duplicates based on a counter object"""
+    counter = getattr(doc._, attr)
+    duplicate_chr = 0
+    for t, c in counter.items():
+        if c > 1:
+            duplicate_chr += len(t) * (c - 1)
+    frac = duplicate_chr / doc._.chr_len
+    return frac
+
+
+def n_gram_counter(doc: Doc, ngram_range: Tuple[int, int]):
+    """Calculate the counts of n-grams in the specified range"""
+    max_len = doc._.len
+    lower, upper = ngram_range
+    shingles_count = defaultdict(lambda: Counter())
+    for i, _ in enumerate(doc):
+        for ngram_size in range(lower, upper + 1):
+            end = i + ngram_size
+            if end < max_len:
+                span = doc[i:end]
+                shingles_count[ngram_size][span.text.lower()] += 1
+    return shingles_count
+
+
+def duplicate_fraction_getter(doc: Doc, attr: str = "lines_counter") -> float:
+    """Calculate the duplicate fraction of based on a counter object
+
+    Args:
+        doc (Doc): A spaCy Doc
+        attr (str, optional): The attribute to the extracted.
+            Defaults to "lines_counter".
+
+    Returns:
+        float: the duplicate fraction
+    """
+    counts = getattr(doc._, attr)
+    n_lines = sum(counts.values())
+    n_unique = len([k for k, c in counts.items() if c == 1])
+    duplicate_fraction = (n_lines - n_unique) / n_lines
+    return duplicate_fraction
+
+
+def set_dynamic_ext(
+    ext_name: str, func: Callable, dynamic_ext_prefix: str = "_", object=Doc
+) -> None:
+    """set a dynamic extension which only computes when required.
```

Comment:
set -> Set

Reply:
Also, I don't understand what this function is doing, from reading the docstring. Can you clarify?

Reply:
Updated the docstring. It is essentially a memoize/cache, where the result is attached to the doc instead of the function.